rfcs/20191127-pip-structure.md

```diff
@@ -0,0 +1,212 @@
+
+#### Circular dependencies
+Estimator depends on TensorFlow. At the same time, TensorFlow includes estimator as a part of its API. This creates a cycle.
+
+![alt_text](https://github.com/annarev/community/blob/pip_structure_rfc/rfcs/20191127-pip-structure/circular_dependency.png "Circular dependency
+between TensorFlow and Estimator.")
+
+#### Metapackage vs base package plans
+[Modular TensorFlow
+RFC](https://github.com/tensorflow/community/blob/master/rfcs/20190305-modular-tensorflow.md) proposes to keep two pip packages:
+tensorflow-base would only contain core TensorFlow (for e.g. no estimator).
+TensorFlow Metapackage would be a thin package defining composition of TensorFlow which includes base, estimator, keras and tensorboard.
+Note that this 2-package approach is not implemented yet. However, its proposal demonstrates how keeping a virtual pip package could be beneficial in the future.
+
+![alt_text](https://github.com/annarev/community/blob/pip_structure_rfc/rfcs/20191127-pip-structure/modular_structure.png "Proposed modular TensorFlow structure.")
+
+Current structure looks more like this (except *tensorflow/* and *tensorflow\_core/* are directories as opposed to separate pip packages) and meant to be the first step towards structure above:
+
+![alt_text](https://github.com/annarev/community/blob/pip_structure_rfc/rfcs/20191127-pip-structure/current_structure.png "Current TensorFlow structure.")
+
+### Current state of IDE code features
+
+#### PyCharm 2019.1.1
+
+* Autocomplete:
+  * Works in most cases after switching to use relative imports.
+  * Doesn’t work for tf.compat.v1.keras and tf.compat.v2.keras.
+  * Doesn’t work for keras if importing it using from import (i.e. `from tensorflow import keras`).
+* Jump-to-definition doesn’t work.
+* Quick documentation doesn’t work.
+
+#### PyCharms with 2019.3 EAP build 193.3793.14
+Latest version of PyCharms added [custom handling for tensorflow](https://github.com/JetBrains/intellij-community/blob/0a08f8212351ee84d602cdc5547f038ce0df79fd/python/src/com/jetbrains/tensorFlow/PyTensorFlow.kt)
+* Autocomplete works in most cases.
+* Doesn’t work for keras if importing it using from import (i.e. `from tensorflow import keras`).
+* Jump-to-definition works.
+* Quick documentation works.
+
+#### VS Code 1.40 (October 2019 release)
+* Autocomplete:
+  * Works in most cases.
+  * Doesn’t work for `tf.estimator` or `tf.keras`.
+  * Doesn’t work for `tf.compat.v1.keras` and `tf.compat.v2.keras`.
+  * Doesn’t work for keras if importing it using from import (i.e. `from tensorflow import keras`).
+* Jump-to-definition doesn’t work.
+* Quick documentation doesn’t work.
```

Comment:
The same holds true for IDEs with [Jedi](https://jedi.readthedocs.io/en/latest/) based completion and introspection (e.g. [`ide-python`](https://github.com/lgeiger/ide-python) for [Atom](https://atom.io/), [`jedi-vim`](https://github.com/davidhalter/jedi-vim), [Spyder](https://www.spyder-ide.org/)).

For more info see https://github.com/davidhalter/jedi/issues/1391

Reply:
I am using Jedi in vim and it is working, by including `tensorflow` into https://jedi.readthedocs.io/en/latest/docs/settings.html#jedi.settings.auto_import_modules -- that way tensorflow is not statically analyzed, bud imported on the Python level, and autocompletion works the same way as in ipython.